kv: panic during shutdown: "Retryable internal error must not happen at this level"

Please answer these questions **for each node in your cluster**:

1. What version of CockroachDB are you using (`cockroach version`)?
`beta-20160101`

2. What operating system and processor architecture are you using?
`Linux c1 4.2.0-34-generic #39-Ubuntu SMP Thu Mar 10 22:13:01 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

3. What flags/environment variables did you pass to `cockroach start`?
`--insecure --alsologtostderr`

Please describe the issue you observed:

1. What did you do?
ctrl-c a running node that was serving queries to photos example.

2. What did you expect to see?
graceful shutdown.

3. What did you see instead?
```
panic: Retryable internal error must not happen at this level

goroutine 23820 [running]:
panic(0x14d07a0, 0xc828add0d0)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).updateState(0xc8202cf1a0, 0x7f20bf1a23b8, 0xc82b893bc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc820e3d238, ...)
        /go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:766 +0xd9f
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).Send(0xc8202cf1a0, 0x7f20bf1a23b8, 0xc82b893bc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc820e3d238, ...)
        /go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:414 +0xddc
github.com/cockroachdb/cockroach/client.(*txnSender).Send(0xc820e3d1e0, 0x7f20bf14e168, 0xc82000d328, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc820e3d238, ...)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:59 +0xb8
github.com/cockroachdb/cockroach/client.(*DB).send(0xc820e3d1e0, 0x12d, 0x0, 0xc82472af60, 0x1, 0x1, 0x0, 0x630d77)
        /go/src/github.com/cockroachdb/cockroach/client/db.go:473 +0x421
github.com/cockroachdb/cockroach/client.(*Txn).send(0xc820e3d1e0, 0x12d, 0xc800000000, 0xc82472af60, 0x1, 0x1, 0x8, 0x64befd)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:649 +0xa46
github.com/cockroachdb/cockroach/client.(*Txn).(github.com/cockroachdb/cockroach/client.send)-fm(0x12d, 0xc800000000, 0xc82472af60, 0x1, 0x1, 0xc82c47c078, 0x9ef7e6)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:361 +0x50
github.com/cockroachdb/cockroach/client.sendAndFill(0xc82c47c090, 0xc82c070380, 0x8, 0x9e8068)
        /go/src/github.com/cockroachdb/cockroach/client/db.go:385 +0x54
github.com/cockroachdb/cockroach/client.(*Txn).RunWithResponse(0xc820e3d1e0, 0xc82c070380, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:361 +0xa9
github.com/cockroachdb/cockroach/client.(*Txn).Run(0xc820e3d1e0, 0xc82c070380, 0xc834975360)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:349 +0x2b
github.com/cockroachdb/cockroach/sql.(*kvFetcher).fetch(0xc82b0f2730, 0xc834974ae0)
        /go/src/github.com/cockroachdb/cockroach/sql/kvfetcher.go:233 +0x3a7
github.com/cockroachdb/cockroach/sql.(*kvFetcher).nextKV(0xc82b0f2730, 0xc829feb1c0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/kvfetcher.go:272 +0x7e
github.com/cockroachdb/cockroach/sql.(*scanNode).nextKey(0xc82b0f2400, 0x1)
        /go/src/github.com/cockroachdb/cockroach/sql/scan.go:124 +0x38
github.com/cockroachdb/cockroach/sql.(*scanNode).Next(0xc82b0f2400, 0x1)
        /go/src/github.com/cockroachdb/cockroach/sql/scan.go:144 +0x70
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc83558ca20, 0xc8ffffffff)
        /go/src/github.com/cockroachdb/cockroach/sql/select.go:108 +0x53
github.com/cockroachdb/cockroach/sql.(*limitNode).Next(0xc82debe7e0, 0xc82017f520)
        /go/src/github.com/cockroachdb/cockroach/sql/limit.go:128 +0x65
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8202b9880, 0x7f20bf1546b8, 0xc82b893680, 0xc8281be078, 0xece8e0999, 0xc82340c061, 0x2638bc0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:894 +0x353
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc8202b9880, 0x7f20bf1546b8, 0xc82b893680, 0xc8281be078, 0x0, 0x1440b08f7cf75653, 0x0, 0xc8281be028, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:778 +0x58d
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8202b9880, 0xc82472a6c0, 0x1, 0x1, 0xc8281be078, 0xc8281be028, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:556 +0x630
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8202b9880, 0xc8281be078, 0x1, 0xc8281be028, 0xc8349749e0, 0xc82472a6c0, 0x1, 0x1, 0xc82c47d2c0, 0xc82c47d2d8, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:488 +0x147
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc820e3d1e0, 0xc8349749e0, 0x2)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:401 +0x325
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc820e3d1e0, 0x1bc0000, 0x0, 0x0, 0xc82c47d418, 0x0)
        /go/src/github.com/cockroachdb/cockroach/client/txn.go:531 +0x410
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8202b9880, 0x7f20bf14e1d8, 0xc82b503b40, 0xc8281be000, 0xc824c54e95, 0x6f, 0x0, 0x0, 0x0, 0xc824c54f00)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:405 +0x892
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8202b9880, 0x7f20bf14e1d8, 0xc82b503b40, 0xc8281be000, 0xc824c54e95, 0x6f, 0xc82472a5c0, 0x1, 0x1, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:295 +0xe2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82445e000, 0x7f20bf14e1d8, 0xc82b503b40, 0xc824c54e95, 0x6f, 0xc82472a5c0, 0x1, 0x1, 0xc82472a5e0, 0x4, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:598 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleExecute(0xc82445e000, 0x7f20bf14e1d8, 0xc82b503b40, 0xc82445e028, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:586 +0x34f
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82445e000, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:291 +0xfce
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc8202e01d8, 0x7f20bf1a2cd0, 0xc82e7c0c60, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:159 +0x929
github.com/cockroachdb/cockroach/server.(*Server).Start.func4.1(0x7f20bf1a2cd0, 0xc82e7c0c60)
        /go/src/github.com/cockroachdb/cockroach/server/server.go:300 +0x4b
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc820064180, 0x7f20bf1a2cd0, 0xc82e7c0c60, 0xc8201fa340)
        /go/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
        /go/src/github.com/cockroachdb/cockroach/util/net.go:127 +0x31f
```

--- a/kv/txn_coord_sender.go
+++ b/kv/txn_coord_sender.go
@@ -791,16 +791,16 @@ func (tc *TxnCoordSender) updateState(
 	case nil:
 		// Nothing to do here, avoid the default case.
 	default:
-		if pErr.GetTxn() != nil {
-			if pErr.CanRetry() {
-				log.Fatalf("retryable internal error must not happen at this level: %s", pErr)
-			} else {
-				// Do not clean up the transaction here since the client might still
-				// want to continue the transaction. For example, a client might
-				// continue its transaction after receiving ConditionFailedError, which
-				// can come from a unique index violation.
-			}
-		}
+		// Do not clean up the transaction here since the client might still
+		// want to continue the transaction. For example, a client might
+		// continue its transaction after receiving ConditionFailedError, which
+		// can come from a unique index violation.
+		//
+		// TODO(bdarnell): Is this valid? Unless there is a single CPut in
+		// the batch, it is difficult to be able to continue after a
+		// ConditionFailedError because it is unclear which parts of the
+		// batch had succeeded on other ranges before one range hit the
+		// failed condition. It may be better to clean up the transaction here.
 	}
 
 	txnID := *newTxn.ID

